engine/execution/ingestion/engine.go

```diff
@@ -180,6 +186,130 @@ func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 	}
 }
 
+// on nodes startup, we need to load all the unexecuted blocks to the execution queues.
+// blocks have to be loaded in the way that the parent has been loaded before loading its children
+func (e *Engine) reloadUnexecutedBlocks() error {
+	err := e.reloadFinalizedUnexecutedBlocks()
+	if err != nil {
+		return fmt.Errorf("could not reload finalized unexecuted blocks")
+	}
+
+	err = e.reloadPendingUnexecutedBlocks()
+	if err != nil {
+		return fmt.Errorf("could not reload pending unexecuted blocks")
+	}
+
+	return nil
+}
+
+func (e *Engine) reloadFinalizedUnexecutedBlocks() error {
+	// get finalized height
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get finalized block: %w", err)
+	}
+
+	// find the first unexecuted and finalized block
+	// we iterate from the last finalized, check if it has been executed,
+	// if not, keep going to the lower height, until we find an executed
+	// block, and then the next height is the first unexecuted.
+	// if there is only one finalized, and it's executed (i.e. genesis),
+	// then the firstUnexecuted is a unfinalized block, which is ok,
+	// because the next loop will ensure it only iterate through finalized
+	// block.
+	firstUnexecuted := final.Height
+	for ; ; firstUnexecuted-- {
+		header, err := e.state.AtHeight(firstUnexecuted).Head()
+		if err != nil {
+			return fmt.Errorf("could not get header at height: %v, %w", firstUnexecuted, err)
+		}
+
+		executed, err := state.IsBlockExecuted(e.unit.Ctx(), e.execState, header.ID())
+		if err != nil {
+			return fmt.Errorf("could not check whether block is executed: %w", err)
+		}
+
+		if executed {
+			firstUnexecuted++
+			break
+		}
+	}
+
+	e.log.Info().Msgf("last finalized and executed height: %v", firstUnexecuted)
+
+	// starting from the first unexecuted block, go through each unexecuted and finalized block
+	// reload its block to execution queues
+	for height := firstUnexecuted; height <= final.Height; height++ {
```

Comment:
We used to fix by starting from the last executed block, which is 1 height lower than the first unexecuted block. 

The reason we needed that fix is that the write operations to update the executed height might be interrupted by a restart, which causes the highest executed height to be inaccurate. Inaccurate means, the 

This new approach no longer depend on the highest executed height. Instead, it goes through each finalized block and pending block, and check if the block has actually been executed before loading them the execution queue.

Reply:
Does this solve the problem of moving finalized height? (finalized block increased while we where running older blocks)

Reply:
Good point. 

I think the finalized height is moving because the follower engine started processing blocks, which triggers BlockProcessable events during the reloading. 

I made a fix to [lock the queue during the reloading](https://github.com/onflow/flow-go/pull/73/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR283-R290)

Reply:
Good catch 👍